core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/ContainerPool.scala

```diff
@@ -342,12 +380,35 @@ object ContainerPool {
   protected[containerpool] def schedule[A](action: ExecutableWhiskAction,
                                            invocationNamespace: EntityName,
                                            idles: Map[A, ContainerData]): Option[(A, ContainerData)] = {
-    idles.find {
-      case (_, WarmedData(_, `invocationNamespace`, `action`, _, activeActivationCount))
-          if activeActivationCount < action.limits.concurrency.maxConcurrent =>
-        true
-      case _ => false
-    }
+    idles
```

Comment:
May be we add a `isMatch` to `ContainerData` like and make it polymorphic
```scala
sealed abstract class ContainerData(val lastUsed: Instant,
                                    val memoryLimit: ByteSize,
                                    val activeActivationCount: Int,
                                    val state: String = "") {
  def isMatch(action: ExecutableWhiskAction, invocationNamespace: EntityName): Boolean = false

  protected def withinConcurrentLimits(action:ExecutableWhiskAction) = 
    activeActivationCount < action.limits.concurrency.maxConcurrent
  }

case class WarmingData(container: Container,
                       invocationNamespace: EntityName,
                       action: ExecutableWhiskAction,
                       override val lastUsed: Instant,
                       override val activeActivationCount: Int = 0)
    extends ContainerData(lastUsed, action.limits.memory.megabytes.MB, activeActivationCount) {
  override def isMatch(action: ExecutableWhiskAction, invocationNamespace: EntityName): Boolean =
    this.action == action && this.invocationNamespace == invocationNamespace && withinConcurrentLimits(action)
}
```

And here do 
```scala
idles.find { case (_, c) => c.isMatch(action, invocationNamespace) }
```

One loss is that we do not prioritize the data types. Probably we can sort the value and pick the head. For e.g. define an `order` field (which captures the hierarchy `WarmingColdData` < `WarmingData` < `WarmedData`) in `ContainerData` and use that for sorting

```scala
idles.find { case (_, c) => c.isMatch(action, invocationNamespace) }.toSeq.sortBy(_.2.order).headOption
```

Reply:
I think `toSeq` in your example will only have the element of the `Option` result of `find()` (single element), so I'm not sure that will work. 
I think `find` is good here, compared to `filter`, but I think having 3 finds with different types is more clear than having an "order". `collect` may work too, but I think will have the same negative impact as `filter`, where every match is collected (while we only want the first).
I reorganized a bit, in any case.

Reply:
is there a comment describing the distinctions between warm, warming, and warming cold and warmed?

Reply:
The `ContainerData` types have docs now: https://github.com/apache/incubator-openwhisk/pull/4186/files#diff-170ae706abad4ff6500ae12be05d0cbbR131